File: moderate/guess_the_number.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	data, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer data.Close()
	scanner := bufio.NewScanner(data)
	scanner.Split(bufio.ScanWords)
	lo, hi := 0, -1
	for scanner.Scan() {
		switch s, c := scanner.Text(), (lo+hi)%2; {
		case hi < 0:
			fmt.Sscan(s, &hi)
		case s == "Lower":
			hi = (lo+hi)/2 + c - 1
		case s == "Higher":
			lo = (lo+hi)/2 + c + 1
		case s == "Yay!":
			fmt.Println((lo+hi)/2 + c)
			lo, hi = 0, -1
		}
	}
}

File: moderate/guess_the_number.rb
File.open(ARGV[0]).each_line do |line|
  s = line.split
  lo = 0
  hi = s[0].to_i
  (1..s.length - 1).each do |x|
    m = lo + hi
    c = m % 2
    case s[x]
    when 'Lower'
      hi = m / 2 + c - 1
    when 'Higher'
      lo = m / 2 + c + 1
    else
      puts m / 2 + c
    end
  end
end
